src/lib/core/error_payload.h

```diff
@@ -0,0 +1,166 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include "trivia/util.h"
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/** Single field of error payload. */
+struct error_field {
+	/** Data size. */
+	uint32_t size;
+	/** MessagePack field value. */
+	char *data;
+	/** Zero terminated field name. */
+	char name[0];
+};
+
+/** An array of key-value pairs. */
```

Comment:
> I think the structure is not about errors at all. It's a general simplest key-value storage. 

I thought the same and even wanted to introduce `struct mp_dict` or something but then decided that it is not generic enough. Because it uses a plain array inside. Quite inefficient when store many key-value pairs in terms of looking for one key. I would better keep it as error-specific struct and if we ever get a worthy implementation of a dictionary - we can drop `error_payload` and transparently replace it with this new struct inside of `struct error`. Anyway the latter wraps payload in its own methods.

Or do you think we still need to move this into a common place even as an array and just optimize it later if will use for anything else?

Reply:
Agree with Vlad. IMO no need to design a general purpose kv storage at this point, when we don't have other users at hand. But even if we did have other potential users, the requirements would likely differ - some would need fast lookup, others wouldn't tolerate frequent reallocations on insertion. Since the `error_payload` implementation is trivial, I think it's okay to duplicate code rather than pursue a general-purpose kv storage that would work for everyone.

Reply:
OK, I think you are right and designing a generic k-v is overcomplicated overkill.